plutus-tx-plugin/executables/profile/Main.hs

```diff
@@ -24,12 +41,117 @@ import qualified PlutusCore.Default        as PLC
 import           Data.Text                 (Text)
+import           Data.Time.Clock           (NominalDiffTime, UTCTime, diffUTCTime)
 import           Prettyprinter.Internal    (pretty)
 import           Prettyprinter.Render.Text (hPutDoc)
 import           System.IO                 (IOMode (WriteMode), withFile)
 
+data Stacks
+  = MkStacks
+  { -- | The variable name.
+    varName           :: String,
+    -- | The time when it starts to be evaluated.
+    startTime         :: UTCTime,
+    -- | The time spent on evaluating the functions it called.
+    timeSpentCalledFn :: NominalDiffTime
+  }
+  deriving (Show)
+
+-- | Write the time log of a program to a file in
+-- the plutus-tx-plugin/executables/profile/ directory.
+writeLogToFile ::
+  ToUPlc a PLC.DefaultUni PLC.DefaultFun =>
+  -- | Name of the file you want to save it as.
+  FilePath ->
+  -- | The program to be profiled.
+  [a] ->
+  IO ()
+writeLogToFile fileName values = do
+  let filePath = "plutus-tx-plugin/executables/profile/"<>fileName<>".timelog"
+  log <- pretty . view _2 <$> (rethrow $ runUPlcProfileExec values)
+  withFile
+    filePath
+    WriteMode
+    (\h -> hPutDoc h log)
+  processed <- processLog filePath
+  writeFile (filePath<>".stacks") processed
+  pure ()
+
+processLog :: FilePath -> IO [Char]
+processLog file = do
+  content <- readFile file
+  -- lEvents is in the form of [[t1,t2,t3,entering/exiting,var]]. Time is chopped to 3 parts.
+  let lEvents =
+        map
+          -- @tail@ strips "[" in the first line and "," in the other lines,
+          -- @words@ turns it to a list of [t1,t2,t3, enter/exit, var]
+          (tail . words)
+          -- turn to a list of events
+          (lines content)
+      lTime = map (unwords . take 3) lEvents
+      -- list of enter/exit
+      lEnterOrExit = map (!! 3) lEvents
+      -- list of var
+      lVar = map (!! 4) lEvents
+      lTripleTimeVar = zip3 (lUTC lTime) lEnterOrExit lVar
+      stacks = getStacks [] lTripleTimeVar
+      fnsStacks = map (intercalate "; " . fst) stacks
+      stacksFgFormat (hdf:tlf) (hdt:tlt)=
+        hdf<>" "<>show hdt<>"\n":stacksFgFormat tlf tlt
+      stacksFgFormat _ _ = []
+  pure $
+    concat $
+      reverse $
+        stacksFgFormat fnsStacks (map ((*1000000) . snd) stacks)
+
+lUTC :: [String] -> [UTCTime]
+lUTC = map (read :: String -> UTCTime)
+
+getStacks ::
+  -- | list of (var, its start time, the amount of time the functions it called spent)
+  [Stacks] ->
+  -- | the input log which is processed to a list of (UTCTime, entering/exiting, var name)
+  [(UTCTime, String, String)] ->
+  -- | a list of (fns it's in, var/function, the time spent on it)
+  [([String],Double)]
+getStacks curStack (hd:tl) =
+  case hd of
+    (time, "entering", var) ->
+      getStacks
+        (MkStacks{varName = var, startTime=time, timeSpentCalledFn = 0 :: NominalDiffTime}:curStack)
+        tl
+    (time, "exiting", var) ->
+      let topOfStack = head curStack
```

Comment:
The general rule is to avoid partial functions if you possibly can, even if you know they're fine (which you do, because you've got another case matching on the empty list). The issue is that you could later rearrange this function to make it not okay, and you wouldn't notice.

And in fact there is a safe way to do this: you can just match on `curStack` in the function definition!